Set up Chrome App shutdown flow.

chuckhays opened the pull request:
Makes the following changes:
* Support both sync/async hangup flow. Sync is used when web browser closes so all cleanup calls can finish before browser closes, async is used when hang up button is pressed. Chrome App can't do sync xhr, so hang up flow has to be async.
* Moves websocket ownership to background.js in the Chrome App
* Adds a queue of cleanup actions in background.js that are executed when the Chrome App window closes

Only background.js knows when the app window closes, and once it has already closed, the window can no longer do xhr or web socket requests. Therefore, background.js has to take care of the clean up calls required to end the in progress call. 

@tkchin @jiayliu 

I'm still working to see if I can unit test more of the code, but wanted to send this out to get the review started, as it is a pretty complex bit of code.

jiayliu commented on src/web_app/js/constants.js:
```diff
@@ -0,0 +1,25 @@
+/* exported Constants */
+'use strict';
+
+var Constants = {
```
Please document what each constant is for.

jiayliu commented on src/web_app/js/background.js:
```diff
@@ -9,14 +9,145 @@
+          data: event
+        });
+      };
+      port.webSocket_ = ws;
```
Encapsulate webSocket_ in Port and provide access methods?

jiayliu commented on src/web_app/js/port.js:
```diff
@@ -0,0 +1,45 @@
+
+'use strict';
+
+var Port = (function() {
```
I thought this is used in background.js, but it appears to be used on the window side. Could you make the name clearer, e.g. WindowPort?

jiayliu commented on src/web_app/js/remotewebsocket.js:
```diff
@@ -0,0 +1,70 @@
+
+'use strict';
+
+var RemoteWebSocket = function(wssUrl) {
```
Please add comments for the purpose of the class.

chuckhays commented:
Updated with feedback.

I deleted SyncPromise like we talked about. After thinking it through, I think it should work fine with regular promises. We just have to be careful if/when we change the hangup method to be sure we don't add anything that cedes control of the thread before it is done.

Ignore the series of commits, I'll squash them before merging anything later.

jiayliu commented on src/web_app/js/windowport.js:
```diff
@@ -0,0 +1,45 @@
+
+'use strict';
+
+var WindowPort = (function() {
```
Please document the purpose of the object.

jiayliu commented on src/web_app/js/util.js:
```diff
@@ -37,21 +37,44 @@ function queryStringToDictionary(queryString) {
+  return sendUrlRequest(method, url, true, body);
+}
+
+function sendSyncUrlRequest(method, url, body) {
```
This seems unused.

chuckhays replied:
It is unused. It was just there as a convenience method. Removed it.

jiayliu commented on Gruntfile.js:
```diff
@@ -133,14 +133,18 @@ module.exports = function(grunt) {
             'src/web_app/js/signalingchannel.js',
             'src/web_app/js/stats.js',
             'src/web_app/js/storage.js',
+            'src/web_app/js/syncpromise.js',
```
remove syncpromise.js

jiayliu commented on src/web_app/js/background.js:
```diff
@@ -9,14 +9,149 @@
+      port.webSocket_ = ws;
+    } else if (message.wsAction === Constants.WS_SEND_ACTION) {
+      trace('RWS: sending: ' + message.data);
+      port.webSocket_.send(message.data);
```
Calling WebSocket methods may throw exceptions and we need to handle.

jiayliu replied:
And need to check websocket state.

chuckhays replied:
I added the POST to wss path like signalingchannel has. We shouldn't get into that state because the readystate is mirrored on the window side, but good to have the code there in case it happens.

jiayliu commented on src/web_app/js/background.js:
```diff
@@ -9,14 +9,149 @@
+    }
+  };
+
+  var createQueuePromise = function(port, message) {
```
nit: rename to createCleanupTaskPromise?

milkypostman commented on src/web_app/js/appcontroller.js:
```diff
@@ -193,7 +193,14 @@ AppController.prototype.showRoomSelection_ = function() {
+    // clean up steps before page is closed.
+    window.onbeforeunload = function() {
+      this.call_.hangup(false);
+    }.bind(this);
```
I think this can be:

```javascript
window.onbeforeunload = this.call_.hangup.bind(this, false);
```

chuckhays replied:
I had it that was, but it was causing odd side effects, possibly because onbeforeunload calls the function with an event object. For some reason it was causing one of those "Are you sure you want to leave this page?" pop ups when you closed the page.

milkypostman replied:
ahh, yes.  because the function is returning something. non-null return value produces a confirmation prompt.

chuckhays replied:
I suspected that, but never verified.

jiayliu commented on src/web_app/js/background.js:
```diff
@@ -9,14 +9,149 @@
+    };
+  };
+
+  var executeQueue = function(port, queue) {
```
nit: rename executeCleanupTasks?

milkypostman commented on src/web_app/js/appcontroller.js:
```diff
@@ -193,7 +193,14 @@ AppController.prototype.showRoomSelection_ = function() {
+    window.onbeforeunload = function() {
+      this.call_.hangup(false);
+    }.bind(this);
+  }
```
Can this change be combined with the other `if` statement below so there is just a single code path where we are considering non chrome app?

jiayliu commented on src/web_app/js/background.js:
```diff
@@ -9,14 +9,149 @@
+  };
+
+  var createQueueCatch = function() {
+    return function(error) {
```
nit: inline the method?

chuckhays replied:
One of our jslint/jscs tools doesn't allow creating methods in a loop, to avoid modified closures. It isn't smart enough to detect that you aren't using any variables from the closure. This is the simplest way to make the rule happy. The other option is to disable the rule for that block.

jiayliu replied:
Cache the function in a local var so we don't create a new closure?

jiayliu commented on src/web_app/js/background.js:
```diff
@@ -9,14 +9,149 @@
+      // Execute the cleanup queue.
+      executeQueue(port, port.queue_).then(function() {
+        // Close web socket.
+        if (port.webSocket_) {
```
Check if the websocket readystate is open?

jiayliu commented on src/web_app/js/background_test.js:
```diff
@@ -0,0 +1,195 @@
+  var ws = this.mockPort_.webSocket_;
+
+  ws.onopen();
+  assertEquals(Constants.WS_ACTION, message.action);
```
Consolidate checking action/wsAction/wsevent into a helper method?

jiayliu commented on src/web_app/js/call.js:
```diff
@@ -89,19 +133,42 @@ Call.prototype.hangup = function() {
+  var path = this.createLeaveUrl_();
+
+  var sendLeave = sendUrlRequest('POST', path, async).catch(function(error) {
+    trace('Error sending /leave: ' + error.message);
```
nit: create helper method to trace error.message?

jiayliu commented on src/web_app/js/call.js:
```diff
@@ -89,19 +133,42 @@ Call.prototype.hangup = function() {
+  }.bind(this));
+
+  var sendBye = sendLeave.then(function() {
+    // Send bye to other client.
```
nit: to the other client.

jiayliu commented on src/web_app/js/common_test.js:
```diff
@@ -0,0 +1,90 @@
+
+'use strict';
+
+var FAKE_WSS_URL = 'wss://foo.com';
```
nit: rename the file to test_mocks.js?

jiayliu commented on src/web_app/js/background.js:
```diff
@@ -9,14 +9,149 @@
+    try {
+      port.postMessage(message);
+    }
+    catch (ex) {
```
nit: catch on the same line as }

jiayliu commented on src/web_app/js/background.js:
```diff
@@ -9,14 +9,149 @@
+      };
+      ws.onerror = function() {
+        trace('RWS: onerror.');
+        sendMessageToWindow(port, {
```
nit: make sendMessageToWindow take arguments |action|, |wsAction|, |wsEvent| and |data|?

jiayliu commented on src/web_app/js/call.js:
```diff
@@ -61,6 +61,45 @@ Call.prototype.start = function(roomId) {
   }
 };
 
+Call.prototype.createCleanupQueue_ = function() {
```
Add test to verify cleanup tasks are queued and removed correctly?

chuckhays commented:
Made most of the changes, working on a unit test for the queue creation, and I still need to do the error handling for the web socket in background.js.

chuckhays commented:
I think I've addressed the comments, just need to decide what we want to do with any exceptions that .send() might throw.

jiayliu commented on src/web_app/js/background.js:
```diff
@@ -9,14 +9,151 @@
+    }
+  };
+
+  var createCleanupTaskPromise = function(port, message) {
```
If you call the inner function executeCleanupTask(port, message), you can just make the caller do 
promise.then(executeCleanupTask.bind(port, message))

chuckhays replied:
Would that have to be .bind(null, port, message) in this case?

jiayliu replied:
Right

chuckhays commented:
Do either of you have an opinion on what to do with exceptions thrown by .send in the websocket owned by background.js? @tkchin @jiayliu

jiayliu commented on src/web_app/js/background.js:
```diff
@@ -9,14 +9,151 @@
+      }
+    } else if (message.wsAction === Constants.WS_CLOSE_ACTION) {
+      trace('RWS: close');
+      port.webSocket_.close();
```
check webSocket_ is not null?

jiayliu commented on src/web_app/js/background.js:
```diff
@@ -9,14 +9,151 @@
+        try {
+          port.webSocket_.send(message.data);
+        } catch (ex) {
+          trace('RWS: exception sending: ' + ex);
```
We probably want to populate the error back to the window

chuckhays replied:
I think that makes sense, right now we don't handle this type of exception in signalingchannel, but we should probably do the same thing in both places.

jiayliu replied:
On Fri, Feb 20, 2015 at 2:51 PM, Chuck Hays <notifications@github.com>
wrote:

> In src/web_app/js/background.js
> <https://github.com/webrtc/apprtc/pull/82#discussion_r25110228>:
>
> > +        trace('RWS: onclose - code: ' + event.code + ' reason: ' +
> > +            event.reason);
> > +        sendWSEventMessageToWindow(port, Constants.WS_EVENT_ONCLOSE, event);
> > +      };
> > +      ws.onmessage = function(event) {
> > +        trace('RWS: onmessage: ' + event.data);
> > +        sendWSEventMessageToWindow(port, Constants.WS_EVENT_ONMESSAGE, event);
> > +      };
> > +      port.webSocket_ = ws;
> > +    } else if (message.wsAction === Constants.WS_SEND_ACTION) {
> > +      trace('RWS: sending: ' + message.data);
> > +      if (port.webSocket_ && port.webSocket_.readyState === WebSocket.OPEN) {
> > +        try {
> > +          port.webSocket_.send(message.data);
> > +        } catch (ex) {
> > +          trace('RWS: exception sending: ' + ex);
>
> I think that makes sense, right now we don't handle this type of exception
> in signalingchannel, but we should probably do the same thing in both
> places.
>
> I'm fine with doing it later if we didn't handle it before your change,
with a bug to track it.


> —
> Reply to this email directly or view it on GitHub
> <https://github.com/webrtc/apprtc/pull/82/files#r25110228>.
>



-- 
Jiayang Liu

jiayliu commented:
I think we need to send the error to the window, so we can update UI, etc

On Fri, Feb 20, 2015 at 2:42 PM, Chuck Hays <notifications@github.com>
wrote:

> Do either of you have an opinion on what to do with exceptions thrown by
> .send in the websocket owned by background.js? @tkchin
> <https://github.com/tkchin> @jiayliu <https://github.com/jiayliu>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/webrtc/apprtc/pull/82#issuecomment-75333168>.
>



-- 
Jiayang Liu

jiayliu commented on src/web_app/js/call.js:
```diff
@@ -89,19 +131,45 @@ Call.prototype.hangup = function() {
+    this.params_.roomId = null;
+    this.params_.clientId = null;
+  }.bind(this));
+
```
For sync mode, check if everything is done and spew error if not?

chuckhays replied:
Could you clarify what you mean? The sync mode only runs when the tab/window is closed.

jiayliu replied:
Yes, it's still helpful to log it

chuckhays replied:
I didn't understand what you mean here, what are you suggesting we log, and what conditions?

jiayliu replied:
if (!async) {
  if (this.params.roomId != null || this.params.clientId != null) {
    trace("ERROR: Room state should be reset for sync hangup")
  }
}

jiayliu commented on src/web_app/js/call_test.js:
```diff
@@ -58,3 +65,36 @@ CallTest.prototype.testRestartInitializesMedia = function() {
   assertTrue(mediaStarted);
 };
+
+CallTest.prototype.testSetUpCleanupQueue = function() {
```
Add test for hangup?

jiayliu commented on src/web_app/js/remotewebsocket.js:
```diff
@@ -0,0 +1,76 @@
+};
+
+RemoteWebSocket.prototype.close = function() {
+  this.readyState = WebSocket.CLOSING;
```
Should check readyState is not closed or closing

jiayliu commented on src/web_app/js/remotewebsocket.js:
```diff
@@ -0,0 +1,76 @@
+};
+
+RemoteWebSocket.prototype.send = function(data) {
+  this.sendMessage_({
```
should check readystate is open

tkchin commented on src/web_app/chrome_app/manifest.json:
```diff
@@ -5,7 +5,7 @@
   "app": {
     "background": {
-      "scripts": ["js/background.js"]
+      "scripts": ["js/apprtc.debug.js", "js/background.js"]
```
Need to look at rest of CL first, but I think it'd be better to have a separate compile target for Chrome build with only the required bits in the background. E.g. apprtc.chrome.debug.js

chuckhays replied:
I thought about doing this, but we would be shipping the same code twice in the chrome app package, because appwindow uses apprtc.debug.js, because it uses the same code as the apprtc on the web.

tkchin replied:
Well. As opposed to loading the same code twice? Because now everything is loaded in two pages when not all of it has to be. Seems to increase memory footprint unnecessarily. Don't have to fix it now, because I think either way package size / memory size is small. But it just felt the same like having a bunch of header files in an impl file that aren't actually being used, especially since they all pollute the global namespace currently.

chuckhays replied:
I'll add an issue to do this. Probably worth it to get things set up correctly sooner vs later.

tkchin commented on src/web_app/js/call.js:
```diff
@@ -89,19 +135,51 @@ Call.prototype.hangup = function() {
+  return setParams;
+};
+
+Call.prototype.createLeaveUrl_ = function() {
```
nit: getLeaveUrl_

tkchin commented on src/web_app/js/signalingchannel.js:
```diff
@@ -122,14 +126,15 @@ SignalingChannel.prototype.close = function() {
+  var path = this.getWssPostUrl();
+
+  return sendUrlRequest('DELETE', path, async).catch(function(error) {
+    trace('Error deleting web socket connection: ' + error.message);
```
Not referencing this in the function, why bind?

milkypostman replied:
We never know what the future holds.

chuckhays replied:
Habit mostly.

tkchin commented on src/web_app/js/util.js:
```diff
@@ -37,21 +37,40 @@ function queryStringToDictionary(queryString) {
     };
-    xhr.open(method, url, true);
+
+    var xhr;
```
nit: declare this before reportResults

tkchin commented on src/web_app/js/util.js:
```diff
@@ -37,21 +37,40 @@ function queryStringToDictionary(queryString) {
+    var xhr;
+    xhr = new XMLHttpRequest();
+    if (async) {
+      xhr.onreadystatechange = function() {
```
onreadystatechange isn't called for sync requests?

chuckhays replied:
Interesting, I had thought it wasn't, but it appears that at least in Chrome and FF it is called even with async=false.

tkchin commented on src/web_app/js/windowport.js:
```diff
@@ -0,0 +1,49 @@
+// It opens a Port object to send and receive messages. When the Chrome
+// App window is closed, background.js receives notification and can
+// handle clean up tasks.
+var WindowPort = (function() {
```
Unusual (to my eyes at least, but not incorrect). Is there reason not to use the common way of defining new object? Why stuff it in a closure instead? If the goal is to have just one port object can wrap just the port in a closure. It looks like you're trying to add namespace here, but we haven't done that elsewhere in the project.

ie. 

var sendPortMessage;
var addPortMessageListener;
(function() {
  var port = chrome.runtime.connect();
  if (!port) { <report error> }
  sendPortMessage = function(message) {...};
  addPortMessageListener = function(listener) {...};
})();

or if multiple calls to port return the same channel:

var WindowPort = function() {
  this.port_ = chrome.runtime.connect();
};

WindowPort.prototype.sendMessage = function(message) ...
etc

chuckhays replied:
I was indeed trying to add a namespace, as we have so many globals already, it seems desirable to at least slow the rate of increase. I'll change it if you want though.

The difference between what I wrote and your first example is only the namespace (as lazy initialization of the port object) as far as I can tell.

For the second example, the actual port object (returned by connect()) should be a singleton. The reason is that on the background side, the port object is what ties the requests to their correct web socket, so the cleanup queue requests need to come to the same port to be able to use the web socket created by RemoteWebSocket's requests.

tkchin replied:
For namespace can we follow this pattern? (I think it's cleaner than repeating declaration at the end of the file. Also see style guide for namespace naming conventions).

var apprtc = apprtc || {};
apprtc.port = apprtc.port || {};
(function() {
  var port_ = chrome.runtime.connect();
  apprtc.port.sendMessage = function(message) {...}
  apprtc.port.addListener = function(listener) {...}
})();

tkchin commented on src/web_app/js/call.js:
```diff
@@ -209,6 +287,9 @@ Call.prototype.connectToRoom_ = function(roomId) {
         .then(function() {
           this.startSignaling_();
+          if (isChromeApp()) {
+            this.createCleanupQueue_();
```
nit: queueCleanupMessages instead, not actually creating something here.
Can you add a comment explaining that we need to queue messages earlier because when window dies only background.js knows?

tkchin commented:
Just minor comments. I think this approach makes sense given our current priorities.

jiayliu commented:
lgtm barring Zeke's comments
Thanks!

chuckhays commented:
@tkchin Your comment about xhr helped me discover that the sync cleanup wasn't actually running completely sync. Even if the promise has completed, .then apparently schedules the body to run on an immediate timeout or something similar. I had to change to another strategy @jiayliu and I had discussed to execute the cleanup steps to ensure they happen synchronously.

PTAL at the most recent change. I'll have to merge after, but there shouldn't be any major conflicts, last time I checked it was just in gruntfile.js.

tkchin commented:
Ok, so I'm guessing that when Promise then block executes async the page is gone since we're in the onbeforeunload handler? [Because promise chaining should be fine, the only sync requirement is that GAE receives /leave before we DELETE off Collider]

chuckhays commented:
I squashed down most of the previous commits (leaving the last couple for now, will squash before merging) and rebased on master.

jiayliu commented on src/web_app/js/background.js:
```diff
@@ -9,14 +9,152 @@
+      var ws = new WebSocket(message.wssUrl);
+      port.wssPostUrl_ = message.wssPostUrl;
+      ws.onopen = function() {
+        trace('RWS: onopen');
```
nit: the trace before sendWSEventMessageToWindow could probably be removed, since sendWSEventMessageToWindow already calls trace.

jiayliu commented:
Could you add tests for Call.hangup()?

On Wed, Feb 25, 2015 at 3:41 PM, Chuck Hays <notifications@github.com>
wrote:

> I squashed down most of the previous commits (leaving the last couple for
> now, will squash before merging) and rebased on master.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/webrtc/apprtc/pull/82#issuecomment-76086355>.
>



-- 
Jiayang Liu

chuckhays commented:
Added unit test and removed verbose logging.

chuckhays commented:
The other problem with testing the async verison is:
https://github.com/webrtc/apprtc/issues/59

Because we use a chain of .then() in the code, the simplified promise implementation we use in the unit test causes only the last .then body to be used.

When I filed that bug, I tried to do a simple implementation to support chaining, but it got too complex, it was basically turning into a full fledged promises polyfill.

jiayliu commented:
Thanks for investigatin. I'm fine to leave it for now.
lgtm

On Fri, Feb 27, 2015 at 1:36 PM, Chuck Hays <notifications@github.com>
wrote:

> The other problem with testing the async verison is:
> #59 <https://github.com/webrtc/apprtc/issues/59>
>
> Because we use a chain of .then() in the code, the simplified promise
> implementation we use in the unit test causes only the last .then body to
> be used.
>
> When I filed that bug, I tried to do a simple implementation to support
> chaining, but it got too complex, it was basically turning into a full
> fledged promises polyfill.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/webrtc/apprtc/pull/82#issuecomment-76476298>.
>



-- 
Jiayang Liu
